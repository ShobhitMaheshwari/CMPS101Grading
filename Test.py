import os, subprocess, shlex
from timeout import timeout
class Test :
	def __init__ (self, testname, path):
		self.path = path
		self.commands = [command.rstrip() for command in open(os.path.join(path, testname))]
		self.url = self.commands[0]
		self.a, self.b = map(lambda x: int(x), self.commands[1].split())
	 	self.points = self.a + self.b
		self.description = self.commands[2]
		self.name = testname
		self.report = ""
		self.score = 0
		del self.commands[0]
		del self.commands[0]
		del self.commands[0]

	@timeout(25)
	def runTest(self):
		FNULL = open(os.devnull, 'w')
		for command in self.commands:
			try :
				testResults = subprocess.check_output(command, shell=True, cwd=self.path)
			except subprocess.CalledProcessError as error :
				self.report +="Test Failure for command "  + command + "\n"
			#Perform diff comparison of expected output and actual output
		try:
			#raise ValueError('A very specific bad thing happened')
			subprocess.check_call(("diff -w -b %s.out %s.expected") % (self.name, self.name), shell=True, cwd = self.path, stdout=FNULL)
			self.score += self.a
		except :
			try:
				#diff <( tr -d ' \n' <Test_Append.out ) <( tr -d ' \n' <Test_Append.expected)
				#print((str(shlex.quote(("diff <( tr -d ' \\n' <%s.out ) <( tr -d ' \\n' <%s.expected)")% (self.name, self.name)))))
				subprocess.check_output((("tr -d '\'' \\n\\t'\'' <%s.out > %s.1.out")%(self.name, self.name)), shell=True, cwd=self.path, executable='/bin/bash')
#				subprocess.check_output((("tr -dc '\''[[:print:]]'\'' <%s2.out > %s1.out")%(self.name, self.name)), shell=True, cwd=self.path, executable='/bin/bash')
				subprocess.check_output((("tr -d '\'' \\n\\t'\'' <%s.expected > %s.1.expected")%(self.name, self.name)), shell=True, cwd=self.path, executable='/bin/bash')
				subprocess.check_call((("diff %s.1.out %s.1.expected")% (self.name, self.name)), shell=True, cwd = self.path, stdout=FNULL)
				#self.report +=("Test failed due to comparison failure on spaces and newlines.") + "\n"
				self.score += self.a
			except:
				if("Test Failure for command" not in self.report):
					self.report +=("Test failed due to comparison failure.") + "\n"
				return

		#valgrind --leak-check=full ./Test_Constructor >> Test_Constructor.out
		with open(os.path.join(self.path, "memoryLeak")) as f:
			content = f.readlines()
			for line in content:
				if 'no leaks are possible' in line:
					self.report += "ok\n"
					self.score += self.b
					return
		self.report +=("Test failed due to memory leak.") + "\n"
