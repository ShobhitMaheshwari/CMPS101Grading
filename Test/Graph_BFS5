https://github.com/ShobhitMaheshwari/CMPS101Grading/blob/master/Test/Graph_BFS5.c
7 1
BFS (graph modified a little - check parent, distance and path, source)
gcc -std=c99 -Wall -c Graph_BFS5.c > w 2>&1
gcc -std=c99 -Wall -c main.c > w  2>&1
gcc -std=c99 -Wall -o Graph_BFS5 main.o Graph_BFS5.o ../List.o ../Graph.o -pthread -lrt &> r
./Graph_BFS5 >  Graph_BFS5.out
valgrind --leak-check=full --log-file="memoryLeak" ./Graph_BFS5
