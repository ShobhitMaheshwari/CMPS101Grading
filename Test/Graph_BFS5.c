#include<stdio.h>
#include<stdlib.h>
#include "../List.h"
#include "../Graph.h"
#include"Test.h"
void main1(){
	//Graph is from CLRS pg 596 Third edition
	//[r,s,t,u,v,w,x,y] = [1,2,3,4,5,6,7,8]
	//BUT there is directed edge from  u to t
	//and no edge between x and u
        Graph a = newGraph(8);

		addEdge(a, 1, 2);//r-s
		addEdge(a, 1, 5);//r-v
		addEdge(a, 2, 6);//s-w
		addEdge(a, 6, 3);//w-t
		addEdge(a, 6, 7);//w-x
		addEdge(a, 3, 7);//t-x
		addArc(a, 4, 3);//u-t //Note the difference
		//addEdge(a, 4, 7);//u-x
		addEdge(a, 4, 8);//u-y
		addEdge(a, 7, 8);//x-y

		BFS(a, 2);
		printf("%d\n", getSource(a));//2
		printf("%d\n", getParent(a, 4));//8
		printf("%d\n", getDist(a, 4));//4
		List l = newList();
		getPath(l, a, 4);
		printList(stdout, l);// 2 6 7 8 4


		BFS(a,4);//again run BFS
		printf("%d\n", getSource(a));//4
		printf("%d\n", getParent(a, 3));//4
		printf("%d\n", getDist(a, 2));//3
		clear(l);
		getPath(l, a, 2);
		printList(stdout, l);//4 3 6 2


		//add edge u-x
		addEdge(a, 4, 7);
		BFS(a,2);//again run BFS

		printf("%d\n", getSource(a));//2
		printf("%d\n", getParent(a, 4));//7
		printf("%d\n", getDist(a, 4));//3
		clear(l);
		getPath(l, a, 4);
		printList(stdout, l);// 2 6 7 4

		freeList(&l);
        freeGraph(&a);
}
