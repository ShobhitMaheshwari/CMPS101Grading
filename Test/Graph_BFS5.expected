2
8
4
2 6 7 8 4
4
4
3
4 3 6 2
2
7
3
2 6 7 4
